url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/CloudletSchedulerTimeShared.java

```diff
@@ -0,0 +1,540 @@
+	 * the actual simulation.
+	 * 
+	 * @pre $none
+	 * @post $none
+	 */
+	public CloudletSchedulerTimeShared() {
+		super();
+		cloudletExecList = new ArrayList<ResCloudlet>();
+		cloudletPausedList = new ArrayList<ResCloudlet>();
+		cloudletFinishedList = new ArrayList<ResCloudlet>();
+		currentCPUs = 0;
+	}
+
+	/**
+	 * Updates the processing of cloudlets running under management of this scheduler.
+	 * 
+	 * @param currentTime current simulation time
+	 * @param mipsShare array with MIPS share of each processor available to the scheduler
+	 * @return time predicted completion time of the earliest finishing cloudlet, or 0 if there is
+	 *         no next events
+	 * @pre currentTime >= 0
+	 * @post $none
+	 */
+	@Override
+	public double updateVmProcessing(double currentTime, List<Double> mipsShare) {
+		setCurrentMipsShare(mipsShare);
+		double timeSpam = currentTime - getPreviousTime();
+
+		for (ResCloudlet rcl : getCloudletExecList()) {
+			rcl.updateCloudletFinishedSoFar((long) (getCapacity(mipsShare) * timeSpam * rcl.getNumberOfPes() * 1000000));
+		}
+
+		if (getCloudletExecList().size() == 0) {
+			setPreviousTime(currentTime);
+			return 0.0;
+		}
+
+		// check finished cloudlets
+		double nextEvent = Double.MAX_VALUE;
+		List<ResCloudlet> toRemove = new ArrayList<ResCloudlet>();
+		for (ResCloudlet rcl : getCloudletExecList()) {
+			long remainingLength = rcl.getRemainingCloudletLength();
+			if (remainingLength == 0) {// finished: remove from the list
+				toRemove.add(rcl);
+				cloudletFinish(rcl);
+				continue;
+			}
+		}
+		getCloudletExecList().removeAll(toRemove);
+
+		// estimate finish time of cloudlets
+		for (ResCloudlet rcl : getCloudletExecList()) {
+			double estimatedFinishTime = currentTime
+					+ (rcl.getRemainingCloudletLength() / (getCapacity(mipsShare) * rcl.getNumberOfPes()));
+			if (estimatedFinishTime - currentTime < 0.1) {
+				estimatedFinishTime = currentTime + 0.1;
+			}
+
+			if (estimatedFinishTime < nextEvent) {
+				nextEvent = estimatedFinishTime;
+			}
+		}
+
+		setPreviousTime(currentTime);
+		return nextEvent;
+	}
+
+	/**
+	 * Gets the capacity.
+	 * 
+	 * @param mipsShare the mips share
+	 * @return the capacity
+	 */
+	private double getCapacity(List<Double> mipsShare) {
+		double capacity = 0.0;
+		int cpus = 0;
+		for (Double mips : mipsShare) {
+			capacity += mips;
+			if (mips > 0.0) {
+				cpus++;
+			}
+		}
+		currentCPUs = cpus;
+
+		int pesInUse = 0;
+		for (ResCloudlet rcl : getCloudletExecList()) {
+			pesInUse += rcl.getNumberOfPes();
+		}
+
+		if (pesInUse > currentCPUs) {
+			capacity /= pesInUse;
+		} else {
+			capacity /= currentCPUs;
+		}
+		return capacity;
+	}
+
+	/**
+	 * Cancels execution of a cloudlet.
+	 * 
+	 * @param cloudletId ID of the cloudlet being cancealed
+	 * @return the canceled cloudlet, $null if not found
+	 * @pre $none
+	 * @post $none
+	 */
+	@Override
+	public Cloudlet cloudletCancel(int cloudletId) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.